**NOTA:**

Artur:

1. análise do conjunto de dados e sua breve descrição  ✅

5. análise preditiva/ aprendizagem de máquina (classificação ou regressão conforme apropriado), comparando o desempenho de vários modelos

Maria:

2. tarefas de pré-processamento dos dados (se julgar necessárias)

6. seleção/ importância das variáveis usadas pelos modelos

Catarina:

3. sumarização dos dados (estatística descritiva, exploração com recurso a gráficos)

4. análise estatística multivariada (não supervisionada): clustering, redução de dimensionalidade/ visualização

# Practical Assignment 2 – Python Application
## 1. Introduction
The goal of this practical assignment is to apply data analysis and knowledge extraction techniques to biological data, with a focus on predicting the mutagenicity of chemical compounds. The task involves developing a complete pipeline in Python, including data loading, preprocessing, exploratory analysis, predictive modeling, and performance evaluation.

The dataset used corresponds to the Ames test, a widely adopted biological assay for identifying compounds with mutagenic potential. It includes a matrix of numerical molecular descriptors (features) and a binary label indicating whether each compound is mutagenic (1) or not (0).

This work provides an opportunity to apply several concepts covered throughout the semester, including multivariate analysis, supervised learning, feature selection, and data visualization.

## 2. Imports

In [2]:
import pandas as pd
import numpy as np
import pprint

## 3. Assigment Context - Ames Mutagenicity

### Dataset Description: 

Mutagenicity means the ability of a drug to induce genetic alterations. Drugs that can cause damage to the DNA can result in cell death or other severe adverse effects. Nowadays, the most widely used assay for testing the mutagenicity of compounds is the Ames experiment which was invented by a professor named Ames. The Ames test is a short-term bacterial reverse mutation assay detecting a large number of compounds which can induce genetic damage and frameshift mutations. The dataset is aggregated from four papers.

### Task Description: 

Binary classification. Given a drug SMILES string, predict whether it is mutagenic `(1)` or not mutagenic `(0)`.

## 4. Data Exploration with Python

In [5]:
# Load the features dataset
features_path = "data_csv\\ames_features.csv"
features_df = pd.read_csv(features_path)

# Basic dataset information
features_info = {
    "Shape": features_df.shape,
    "Columns": features_df.columns.tolist(),
    "Missing Values": features_df.isnull().sum().sum(),
    "Preview": features_df.head()
}

print("Dataset Shape:")
pprint.pprint(features_info["Shape"])

Dataset Shape:
(7273, 208)


In [6]:
print("\n Feature Columns (truncated list):")
pprint.pprint(features_info["Columns"])


 Feature Columns (truncated list):
['MaxEStateIndex',
 'MinEStateIndex',
 'MaxAbsEStateIndex',
 'MinAbsEStateIndex',
 'qed',
 'MolWt',
 'HeavyAtomMolWt',
 'ExactMolWt',
 'NumValenceElectrons',
 'NumRadicalElectrons',
 'MaxPartialCharge',
 'MinPartialCharge',
 'MaxAbsPartialCharge',
 'MinAbsPartialCharge',
 'FpDensityMorgan1',
 'FpDensityMorgan2',
 'FpDensityMorgan3',
 'BCUT2D_MWHI',
 'BCUT2D_MWLOW',
 'BCUT2D_CHGHI',
 'BCUT2D_CHGLO',
 'BCUT2D_LOGPHI',
 'BCUT2D_LOGPLOW',
 'BCUT2D_MRHI',
 'BCUT2D_MRLOW',
 'BalabanJ',
 'BertzCT',
 'Chi0',
 'Chi0n',
 'Chi0v',
 'Chi1',
 'Chi1n',
 'Chi1v',
 'Chi2n',
 'Chi2v',
 'Chi3n',
 'Chi3v',
 'Chi4n',
 'Chi4v',
 'HallKierAlpha',
 'Ipc',
 'Kappa1',
 'Kappa2',
 'Kappa3',
 'LabuteASA',
 'PEOE_VSA1',
 'PEOE_VSA10',
 'PEOE_VSA11',
 'PEOE_VSA12',
 'PEOE_VSA13',
 'PEOE_VSA14',
 'PEOE_VSA2',
 'PEOE_VSA3',
 'PEOE_VSA4',
 'PEOE_VSA5',
 'PEOE_VSA6',
 'PEOE_VSA7',
 'PEOE_VSA8',
 'PEOE_VSA9',
 'SMR_VSA1',
 'SMR_VSA10',
 'SMR_VSA2',
 'SMR_VSA3',
 'SMR_VSA4',
 'SMR_VSA

In [71]:
print("Total Missing Values:")
print(features_info["Missing Values"])

Total Missing Values:
0


In [72]:
print("Dataset Preview (first 5 rows):")
features_info["Preview"]

Dataset Preview (first 5 rows):


,MaxEStateIndex,MinEStateIndex,MaxAbsEStateIndex,MinAbsEStateIndex,qed,MolWt,HeavyAtomMolWt,ExactMolWt,NumValenceElectrons,NumRadicalElectrons,...,fr_sulfide,fr_sulfonamd,fr_sulfone,fr_term_acetylene,fr_tetrazole,fr_thiazole,fr_thiocyan,fr_thiophene,fr_unbrch_alkane,fr_urea
0,11.663401,-0.396099,11.663401,0.008895,0.181963,342.310,332.230,342.06406,124.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
1,11.824743,-0.170602,11.824743,0.170602,0.269289,301.345,286.225,301.11030,112.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
2,14.431072,-0.378575,14.431072,0.128988,0.184582,646.614,628.470,646.11650,232.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
3,10.422196,-0.670278,10.422196,0.248657,0.103849,157.133,150.077,157.05997,60.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0
4,10.521759,-0.752407,10.521759,0.278148,0.342148,138.086,136.070,138.01778,50.0,0.0,...,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,0.0,1.0


### **Output**

In [73]:
# Load the features dataset
features_path = "/home/arog/Documents/GitHub/ECDB_Project/trabalho_py/data_csv/ames_outputs.csv"
features_df = pd.read_csv(features_path)

# Basic dataset information
features_info = {
    "Shape": features_df.shape,
    "Columns": features_df.columns.tolist(),
    "Missing Values": features_df.isnull().sum().sum(),
    "Preview": features_df.head(),
    "Class Distribution": features_df['Y'].value_counts().to_dict()
}

print("Dataset Shape:")
pprint.pprint(features_info["Shape"])

Dataset Shape:
(7273, 3)


In [74]:
print("Feature Columns (truncated list):")
pprint.pprint(features_info["Columns"])

Feature Columns (truncated list):
['ids', 'smiles', 'Y']


In [75]:
print("Total Missing Values:")
print(features_info["Missing Values"])

Total Missing Values:
0


In [76]:
print("Dataset Preview (first 5 rows):")
features_info["Preview"]

Dataset Preview (first 5 rows):


,ids,smiles,Y
0,Drug 0,O=[N+]([O-])c1ccc2ccc3ccc([N+](=O)[O-])c4c5ccc...,1
1,Drug 1,O=[N+]([O-])c1c2c(c3ccc4cccc5ccc1c3c45)CCCC2,1
2,Drug 2,O=c1c2ccccc2c(=O)c2c1ccc1c2[nH]c2c3c(=O)c4cccc...,0
3,Drug 3,[N-]=[N+]=CC(=O)NCC(=O)NN,1
4,Drug 4,[N-]=[N+]=C1C=NC(=O)NC1=O,1


In [81]:
print("Class Distribution — 1 = mutagenic, 0 = non-mutagenic:")
pprint.pprint(features_info["Class Distribution"])

Class Distribution — 1 = mutagenic, 0 = non-mutagenic:
{0: 3299, 1: 3974}


## 4. Datasets Overview

This project utilizes two datasets derived from the Ames mutagenicity test: one containing molecular features extracted from chemical structures, and another providing binary mutagenicity classification labels.

---

### Dataset Features

This dataset includes computed molecular descriptors and fragment-based features for each compound, derived from its SMILES (Simplified Molecular Input Line Entry System) representation.

- **Rows (samples)**: 7,273 compounds  
- **Columns (features)**: 208 numerical descriptors  
- **Types of features**:
  - *Physicochemical*: `MolWt`, `TPSA`, `LogP`, `qed`, etc.
  - *Topological and electronic*: `Chi`, `BalabanJ`, `EState`, etc.
  - *Structural fragments (binary flags)*: `fr_nitro`, `fr_urea`, `fr_halogen`, etc.

The dataset contains no missing values and is ready for use in machine learning tasks.

---

### Dataset Output

This dataset contains binary classification labels indicating the mutagenicity of each compound, aligned with the compounds in the feature dataset.

- **Rows**: 7,273 (one per compound)
- **Columns**:
  - `ids`: Unique compound identifier
  - `smiles`: SMILES representation of the compound
  - `Y`: Binary label for mutagenicity  
    - `1` = Mutagenic  
    - `0` = Non-mutagenic

#### Class Distribution

| Label             | Count |
|-------------------|--------|
| Mutagenic (1)     | 3,974  |
| Non-mutagenic (0) | 3,299  |

There are no missing values in the output dataset. The class distribution is slightly imbalanced, with more mutagenic than non-mutagenic samples.

---

Together, these datasets form a complete and clean structure for building supervised learning models to predict the mutagenic potential of chemical compounds.


# Supervised ML